src/nupic/utils/Random.hpp

```diff
@@ -76,155 +72,104 @@ typedef UInt64 (*RandomSeedFuncPtr)();
  * reproducibility in the case that a test failure is triggered by a particular
  * seed.
  *
- * Random should not be used if cryptographic strength is required (e.g. for
- * generating a challenge in an authentication scheme).
- *
  * @todo Add ability to specify different rng algorithms.
+ *
+ * API:
+ * there are 2 main functions,
+ * getReal()
+ * getUInt()
+ *
  */
-class RandomImpl {
-public:
-  RandomImpl(UInt64 seed);
-  ~RandomImpl(){};
-  UInt32 getUInt32();
-  bool operator==(const RandomImpl &o) const;
-  inline bool operator!=(const RandomImpl &other) const {
-    return !operator==(other);
-  }
-  // Note: copy constructor and operator= are needed
-  // The default is ok.
-private:
-  friend std::ostream &operator<<(std::ostream &outStream, const RandomImpl &r);
-  friend std::istream &operator>>(std::istream &inStream, RandomImpl &r);
-  static const UInt32 VERSION = 2;
-  // internal state
-  static const int stateSize_ = 31;
-  static const int sep_ = 3;
-  UInt32 state_[stateSize_];
-  int rptr_;
-  int fptr_;
-};
-
-
-
-
-class Random : public Serializable {
+class Random : public Serializable  {
 public:
-  /**
-   * Retrieve the seeder. If seeder not set, allocates the
-   * singleton and and initializes the seeder.
-   */
-  static RandomSeedFuncPtr getSeeder();
-
   Random(UInt64 seed = 0);
 
-  // support copy constructor and operator= -- these require non-default
-  // implementations because of the impl_ pointer.
-  // They do a deep copy of impl_ so that an RNG and its copy generate the
-  // same set of numbers.
-  Random(const Random &);
-  Random &operator=(const Random &);
-  ~Random();
-
   // save and load serialized data
-  virtual void save(std::ostream &stream) const override { stream << *this; }
-  virtual void load(std::istream &stream) override { stream >> *this; }
-  virtual void saveToFile(std::string filePath) const override { Serializable::saveToFile(filePath); }
-  virtual void loadFromFile(std::string filePath) override { Serializable::loadFromFile(filePath); }
+  void save(std::ostream &stream) const override { stream << *this; }
+  void load(std::istream &stream) override { stream >> *this; }
+  void saveToFile(std::string filePath) const override { Serializable::saveToFile(filePath); }
+  void loadFromFile(std::string filePath) override { Serializable::loadFromFile(filePath); }
 
 
   bool operator==(const Random &other) const;
   inline bool operator!=(const Random &other) const {
     return !operator==(other);
   }
 
+  //main API methods:
   // return a value uniformly distributed between 0 and max-1
-  UInt32 getUInt32(UInt32 max = MAX32);
-  UInt64 getUInt64(UInt64 max = MAX64);
+  inline UInt32 getUInt32(const UInt32 max = MAX32) {
+    NTA_ASSERT(max > 0);
+    return dist_uint_32(gen) % max; //TODO the modulo % is not ideal, better use w/o param here, and supply max 
```

Comment:
Modulo is really a division operation and subtraction. Not really all that slow.  Having the max range is important. I remember seeing it actually used.  So if we remove it here we would have to add the modulo to every place it is used.  I agree that modulo of small numbers does not give a good distribution but normally the range is not that small.

Reply:
> So if we remove it here we would have to add the modulo to every place it is used. 

the range is important..the suggestion was to use: 

```
distr = UniformDist<Uint>(0, maxRange) //you can change the range only here
distr.getUInt() //but not here, as is possible currently (is such behavior required?) 
```
instead of `distr(maxUInt).getUInt() % maxRange` 
in which case the distribution takes care of the range as well as of the uniform distribution

ronja

Reply:
Oh, I misunderstood your proposal.  I thought you wanted to remove the range from Random::getUInt64(max) ...etc.   So, yes I agree with you.